### PR TITLE
[Prometheus] Added dynamic_<dataset|namespace> settings to prometheus datasets

### PR DESCRIPTION
**Pull request by Alphayeeeet**
Please label as Enhancement

This PR allows using prometheus datasets as a generic metric datasets in  e.g. Kubernetes environments, as it adds the permissions to write to metrics-*-*. This allows the configuration of annotation based rerouting in ingest pipelines like it is used in kubernetes.container_logs.

If needed, we can add a default rerouting mechanism, based on the kubernetes annotations like it is already provided in kubernetes.container_logs.

If necessary, we can add a generic metric dataset with this setting to the Kubernetes integration instead of using the Prometheus package for that.

### CODE REVIEW

**Comment by Alphayeeeet** [comment]
@andrewkroh or anyone else: Could you please run the CI tests? Thank you

**Comment by Alphayeeeet** [comment]
@andrewkroh Could you please run the CI again? Also is there anyone from elastic-obs-infra-obs I can tag, that may review this PR (as it is open quite a while now and only introduces a small quality of life improvement)?

**Comment by Alphayeeeet** [comment]
Kindly asking for review, because this PR is open quite a while now.  @elastic/obs-infraobs-integrations

**Comment by Alphayeeeet** [comment]
@muthu-mps Im just tagging you as you reviewed my other PR. Could you review this one aswell or is this the reponsibility of a different member of infraobs-integrations? Thank you in advance

**Comment by muthu-mps** [comment]
> @muthu-mps Im just tagging you as you reviewed my other PR. Could you review this one aswell or is this the reponsibility of a different member of infraobs-integrations? Thank you in advance

Thanks @Alphayeeeet. I am tagging @ishleenk17 to take a look into this PR.

**Comment by Alphayeeeet** [comment]
Hi @ishleenk17 just tagging you again to remind for this PR. We are currently blocked in our internal implementation by the rerouting feature. Thank you

File: packages/prometheus/data_stream/collector/manifest.yml [hunk 1]
```diff
@@ -2,6 +2,9 @@ title: Prometheus collector metrics
 type: metrics
 elasticsearch:
   index_mode: "time_series"
+  # Ensures agents have permissions to write data to `metrics-*-*`
+  dynamic_dataset: true
```
Comment (ishleenk17) [inline packages/prometheus/data_stream/collector/manifest.yml:6]: Any specific reason to limit this to collector only ?

File: packages/prometheus/changelog.yml [hunk 2]
```diff
@@ -1,4 +1,9 @@
 # newer versions go on top
+- version: "1.19.0"
+  changes:
+    - description: Allow automatic rerouting for prometheus.collector dataset. This allows generic rerouting of metrics in e.g. Kubernetes environments.
```
Comment (ishleenk17) [inline packages/prometheus/changelog.yml:4]:
```suggestion
    - description: Allow automatic rerouting for collector dataset.  E.g. Routing to Kubernetes environments.
```

**Comment by Alphayeeeet** [comment]
@ishleenk17 If this change could also be useful for the other datasets, I can implement them aswell. But in our currently used structure, only the collector is used as we reroute based on Kubernetes annotations (like already possible with the kubernetes.container_logs dataset of the Kubernetes integration).

**Comment by Alphayeeeet** [comment]
@ishleenk17 @tetianakravchenko @MichaelKatsoulis any updates on this PR by now?

**Comment by ishleenk17** [comment]
@tetianakravchenko : Please check from remote write perspective.
Collector looks good.

**Comment by Alphayeeeet** [comment]
@ishleenk17 & @tetianakravchenko Could you please review and approve the PR if possible? Still waiting for this change and we are still blocked in our metrics implementation...

**Review by ishleenk17 — APPROVED** [review]

**Comment by ishleenk17** [comment]
@elastic/obs-ds-hosted-services : Please check this.

**Review by MichaelKatsoulis — APPROVED** [review]

**Comment by ishleenk17** [comment]
> @tetianakravchenko any updates on this PR?

You have all the needed approvals. We can go ahead and merge it now.

**Comment by Alphayeeeet** [comment]
@ishleenk17 Thanks. unfortunately the auto merge did not work, and im not permitted to merge in general. Thank you anyway

**Pull request merged**